messaging/src/main/java/org/axonframework/tracing/NestingSpanFactory.java

```diff
@@ -0,0 +1,176 @@
+    }
+
+    @Override
+    public void registerSpanAttributeProvider(SpanAttributesProvider provider) {
+        delegateSpanFactory.registerSpanAttributeProvider(provider);
+    }
+
+    @Override
+    public <M extends Message<?>> M propagateContext(M message) {
+        return delegateSpanFactory.propagateContext(message);
+    }
+
+
+    /**
+     * Creates a builder that will create a {@link NestingSpanFactory}.
+     * <p>
+     * Requires the delegate {@link SpanFactory} to be configured.
+     * <p>
+     * The {@link Clock} defaults to the system utc time and the timeLimit {@link Duration} defaults to two minutes.
+     */
+    public static class Builder {
+
+        private SpanFactory delegateSpanFactory;
+        private Duration timeLimit = Duration.ofMinutes(2);
+        private Clock clock = Clock.systemUTC();
+
+        /**
+         * Defines the delegate {@link SpanFactory} to use, which actually provides the spans.
+         *
+         * @param spanFactory The {@link SpanFactory} to configure for use.
+         * @return The current Builder instance, for fluent interfacing.
+         */
+        public Builder delegate(SpanFactory spanFactory) {
+            BuilderUtils.assertNonNull(spanFactory, "The spanFactory should not be null");
+            this.delegateSpanFactory = spanFactory;
+            return this;
+        }
+
+        /**
+         * Configures the {@link Duration} since original publishing of an event during which it should be considered a
+         * nested span. After that duration, it will become its own separate trace.
+         *
+         * @param timeLimit The amount of time before handling of an event should be considered a separate trace.
+         * @return The current Builder instance, for fluent interfacing.
+         */
+        public Builder timeLimit(Duration timeLimit) {
+            BuilderUtils.assertNonNull(timeLimit, "The timeLimit should not be null");
+            this.timeLimit = timeLimit;
+            return this;
+        }
+
+        /**
+         * Configures the {@link Clock} to use when determining the time passed since publication of an event and the
+         * current time.
+         *
+         * @param clock The {@link Clock} to use when determining the time passed since publication of an event and the
+         *              current time.
+         * @return The current Builder instance, for fluent interfacing.
+         */
+        public Builder clock(Clock clock) {
+            BuilderUtils.assertNonNull(clock, "The clock should not be null");
+            this.clock = clock;
+            return this;
+        }
+
+        /**
+         * Executes the builder's configuration, creating the {@link NestingSpanFactory}.
+         *
+         * @return The span factory.
+         */
+        public NestingSpanFactory build() {
+            BuilderUtils.assertNonNull(delegateSpanFactory, "The delegateSpanFactory should be configured");
```

Comment:
Ah, so here's the "validation."
The reasoning it's typically moved to a dedicated method, is so that implementations of a class can override the Builder (required as per the constructor) and thus also override the validation behavior.